File: app/src/main/java/com/sagar/watchnext/activities/main/home/HomeFragmentViewModelFactory.java
package com.sagar.watchnext.activities.main.home;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.sagar.watchnext.network.repo.TMDBRepository;
import com.sagar.watchnext.network.repo.TmdbMovieRepo;
import com.sagar.watchnext.network.repo.TmdbTvRepo;

import javax.inject.Inject;

/**
 * Created by SAGAR MAHOBIA on 16-Feb-19. at 01:52
 */

@HomeFragmentScope
public class HomeFragmentViewModelFactory implements ViewModelProvider.Factory {

    private TmdbMovieRepo movieRepo;
    private TmdbTvRepo tvRepo;
    private TMDBRepository tmdbRepository;

    @Inject
    HomeFragmentViewModelFactory(TmdbMovieRepo movieRepo, TmdbTvRepo tvRepo, TMDBRepository tmdbRepository) {
        this.movieRepo = movieRepo;
        this.tvRepo = tvRepo;
        this.tmdbRepository = tmdbRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(HomeFragmentViewModel.class)) {
            return (T) new HomeFragmentViewModel(movieRepo, tvRepo,tmdbRepository);
        } else {
            throw new IllegalStateException("unknown view model");
        }
    }

}

File: app/src/main/java/com/sagar/watchnext/activities/main/home/HomeFragmentViewModelFactory.kt
package com.sagar.watchnext.activities.main.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.sagar.watchnext.network.repo.TMDBRepository
import com.sagar.watchnext.network.repo.TmdbTvRepo
import javax.inject.Inject

/**
 * Created by SAGAR MAHOBIA on 16-Feb-19. at 01:52
 */
@HomeFragmentScope
class HomeFragmentViewModelFactory @Inject internal constructor(private val tmdbRepository: TMDBRepository, private val tmdbTvRepo: TmdbTvRepo) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {

        if (modelClass.isAssignableFrom(HomeFragmentViewModel::class.java)) {
            return HomeFragmentViewModel(tmdbRepository, tmdbTvRepo) as T
        } else {
            throw IllegalStateException("unknown view model")
        }
    }

}


/*
POSSIBLE JAVA EQUIVALENT

public class MoviesFragmentViewModelFactory implements ViewModelProvider.Factory {

    private TmdbMovieRepo movieRepo;
    private TMDBRepository tmdbRepository;

    @Inject
    MoviesFragmentViewModelFactory(TmdbMovieRepo movieRepo, TMDBRepository tmdbRepository) {
        this.movieRepo = movieRepo;
        this.tmdbRepository = tmdbRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(MoviesFragmentViewModel.class)) {
            return (T) new MoviesFragmentViewModel(tmdbRepository);
        } else {
            throw new IllegalStateException("unknown view model");
        }
    }
}*/